### PR TITLE
feat(pruner): storage history

### LINKED ISSUE #3694: Respect `Storage History` prune part in the pruner
### Describe the feature

Prune `StorageChangeSet` and `StorageHistoryIndex` tables in the pruner.

If `Account History` prune target is less than 64 blocks, don't prune `StorageChangeSet` table until the target is > 64 blocks. `StorageHistoryIndex` can be safely pruned, as they are not needed for the unwind.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by shekhirin**
Resolves #3694 

Prunes:
1. `StorageChangeSet` table: all keys starting with block numbers (and their subkeys which are storage keys) which need to be pruned
2. `StorageHistory` table:
   1. If the shard has `highest_block_number` less than or equal to the target block number for pruning, delete the shard completely.
   2. If the shard has `highest_block_number` greater than the target block number for pruning, filter block numbers inside the shard which are less than the target block number for pruning.

### CODE REVIEW

File: crates/prune/src/pruner.rs [hunk 1]
```diff
@@ -425,6 +433,112 @@ impl<DB: Database> Pruner<DB> {
...
+                    .skip_while(|block| *block <= to_block as usize)
+                    .collect::<Vec<_>>();
+                if blocks.is_empty() {
+                    // If there are no more blocks in this shard, we need to remove it, as empty
+                    // shards are not allowed.
+                    if key.sharded_key.highest_block_number == u64::MAX {
+                        // If current shard is the last shard for this address and storage slot,
+                        // replace it with the previous shard.
+                        if let Some((prev_key, prev_value)) = cursor.prev()? {
+                            if prev_key.sharded_key.key == key.sharded_key.key {
```
Comment (joshieDo) [inline crates/prune/src/pruner.rs:498]: shouldn't we check `prev_key.address == key.address` as well?

**Review by joshieDo — CHANGES_REQUESTED** [review]

File: crates/prune/src/pruner.rs [hunk 2]
```diff
@@ -446,6 +454,120 @@ impl<DB: Database> Pruner<DB> {
...
+                    // Shard contains only block numbers up to the target one, so we can skip to the
+                    // next storage slot for this address. It is guaranteed that further shards for
+                    // this address and storage slot will not contain the target block number, as
+                    // it's in this shard.
+                    cursor.seek_exact(StorageShardedKey::last(key.address, key.sharded_key.key))?;
+                }
+            } else {
+                // Shard contains block numbers that are higher than the target one, so we need to
+                // filter it. It is guaranteed that further shards for this address and storage slot
+                // will not contain the target block number, as it's in this shard.
```
Comment (mattsse) [inline crates/prune/src/pruner.rs:513]: this is a very useful comment

File: crates/prune/src/pruner.rs [hunk 3]
```diff
@@ -446,6 +454,120 @@ impl<DB: Database> Pruner<DB> {
...
+                let blocks = blocks
+                    .iter(0)
+                    .skip_while(|block| *block <= to_block as usize)
+                    .collect::<Vec<_>>();
```
Comment (mattsse) [inline crates/prune/src/pruner.rs:517]: these are only numbers and not the actual blocks, so we can safely load them

File: crates/prune/src/pruner.rs [hunk 4]
```diff
@@ -446,6 +454,120 @@ impl<DB: Database> Pruner<DB> {
...
+                if blocks.is_empty() {
+                    // If there are no more blocks in this shard, we need to remove it, as empty
+                    // shards are not allowed.
+                    if key.sharded_key.highest_block_number == u64::MAX {
+                        // If current shard is the last shard for this address and storage slot,
+                        // replace it with the previous shard.
+                        if let Some((prev_key, prev_value)) = cursor.prev()? {
+                            if prev_key.address == key.address &&
```
Comment (mattsse) [inline crates/prune/src/pruner.rs:525]: even if we have a bunch of levels, this is still readable imo

**Review by mattsse — APPROVED** [review]
lgtm

**Review by joshieDo — APPROVED** [review]

**Pull request merged**